README.md
# Screenshot

[![API](https://img.shields.io/badge/API-14%2B-orange.svg?style=flat)](https://android-arsenal.com/api?level=14)
[![License Apache 2.0](https://img.shields.io/badge/License-Apache%202.0-blue.svg?style=true)](http://www.apache.org/licenses/LICENSE-2.0)
[![Total Downloads](https://img.shields.io/github/downloads/NmdOffical/Screenshot/total.svg)]()
[![](https://jitpack.io/v/NmdOfficial/Screenshot.svg)](https://jitpack.io/#NmdOfficial/Screenshot)

A new and nice working android library to take screenshots which provides many functions.<br>
If enabled you will see a screenshot preview like you know it from the official android screenshot option.<br>
While the screenshot is being taken, you will also hear the shutter click sound for android 4.1+ devices.<br>


## Usage
Add a dependency to your build.gradle file:
```java
allprojects {
    repositories {
	    ...
	    maven { url 'https://jitpack.io' }
    }
}

dependencies {
    implementation 'com.github.NmdOfficial:Screenshot:-SNAPSHOT'
}
```

```java
import com.nmd.android.support.Screenshot;
```

| Public constructors |
| --- |
| `Screenshot(Context context)` |


## Needed app permissions
```java
android.permission.READ_EXTERNAL_STORAGE;
android.permission.WRITE_EXTERNAL_STORAGE;
```
AndroidMainfest.xml
```java
<manifest xlmns:android...>
  ...
    <uses-permission android:name="android.permission.READ_EXTERNAL_STORAGE" />
    <uses-permission android:name="android.permission.WRITE_EXTERNAL_STORAGE" />
  <application ...
</manifest>
```

Don't forget to set MIN and TARGET SDK values:
```java
<uses-sdk android:minSdkVersion="14" android:targetSdkVersion="xx"/>
```

## Interface

| type | name |
| --- | --- |
| `interface` | OnResultListener<br>*Interface definition for a callback to be invoked when a screenshot was created.* |

Public methods

| type | function |
| --- | --- |
| `abstract void` | result(boolean success, String filePath, Bitmap bitmap)<br>*Called when a screenshot was created.* |

```java
.setCallback(new Screenshot.OnResultListener() {
    @Override
    public void result(boolean success, String filePath, Bitmap bitmap) {
      // your code here.        
    }
});
```

## Public methods

| type | function | description | default value | min api |
| --- | --- | --- | --- | --- |
| `void` | TakeScreenshot() | *Take a screenshot of the current visible screen.* | - |  14 |
| `void` | TakeScreenshotFromView() | *Take a screenshot of any visible view.* | - |  14 |
| `void` | setFileName(String name) | *The filename for the taken screenshot.* | \"Screenshot.png\" |  14 |
| `string` | getFileName() | *Returns the given screenshot filename.* | - |  14 |
| `void` | ShowPreview(boolean enabled) | *If enabled you will see a short preview animation after the screenshot is taken.* | true |  14 |
| `boolean` | ShowPreview() | *Returns true/false if* `ShowPreview(...)` *is enabled/disabled.* | - |  14 |
| `void` | ShowNotification(boolean enabled) | *If enabled you will see a notification in the statusbar after the screenshot is taken.* | true |  14 |
| `boolean` | ShowNotification() | *Returns true/false if* `ShowNotification(...)` *is enabled/disabled.* | - |  14 |
| `void` | NotificationTitle(String name) | *The title text for the notification.* | \"Screenshot..\" |  14 |
| `string` | NotificationTitle() | *Returns the given notification title text.* | - |  14 |
| `void` | NotificationShareTitle(String name) | *The share button text for the notification if* `Show Notification(...)` *is enabled.* | "Share" |   16 |
| `string` | NotificationShareTitle() | *Returns the given notification share title text.* | - |  16 |
| `void` | NotificationBigStyle(boolean enabled) | *If enabled you will see a notification with \"big style\" in the statusbar after the screenshot is taken.* | true |  16 |
| `boolean` | NotificationBigStyle() | *Returns true/false if* `NotificationBigStyle(...)` *is enabled/disabled.* | - |  16 |
| `void` | NotificationShareButton(boolean enabled) | *If enabled you will see a notification with a share button after the screenshot is taken.* | true |  16 |
| `boolean` | NotificationShareButton() | *Returns true/false if* `NotificationShareButton(...)` *is enabled/disabled.* | - |  16 |
| `void` | AllowScreenshots(boolean enabled) | *This feature allows users of your app to make or ban screenshots of their app.<br>If disabled and a person tries to make a screenshot, they will receive then a default system message that this is not possible.* | - |  1 |
| `void` | setDimAmount(float amount) | *Set the amount of dim behind the preview window if* `ShowPreview(...)` *is enabled. Use '0.0' for no dim and '1.0' for full dim.* | 0.5f |  14 |
| `float` | getDimAmount() | *Returns the amount from `setDimAmount(...)` .* | - |  14 |
| `boolean` | arePermissionsGranted() | *Returns true if the write and read permission is granted, else false.* | - |  1 |
| `boolean` | isReadPermissionGranted() | *Returns true if the read permission is granted, else false.* | - |  1 |
| `boolean` | isWritePermissionGranted() | *Returns true if the write permission is granted, else false.* | - |  1 |

*You do not need to add SDK version checks by yourself. This does the library for you.*

## Exampe code

```java
package com.YOURNAME.YOURNAME;

import android.app.Activity;
import android.content.Context;
import android.widget.TextView;
import android.graphics.Color;
import com.nmd.android.support.Screenshot;
...

public class YourClass extends Activity {
  private Context context;
  private TextView txt;
  private Screenshot screenshot;

    public YourClass (...) {
      this.context = getContext();
      Initialize();
    }
    
    private void Initialize() {
      txt = (TextView)findViewById(R.id.editText);
      screenshot = new Screenshot(this.context);
    }

    public void Take() {
      screenshot.NotificationTitle("My screenshot title");
      screenshot.setCallback(new Screenshot.OnResultListener() {
        @Override
        public void result(boolean success, String filePath, Bitmap bitmap) {
          txt.setText(filePath);
          txt.setTextColor(success ? Color.GREEN : Color.RED);
          // if success is true then set text color to green, else red
        }
      });
      //After you have done your settings let's take the screenshot
      screenshot.TakeScreenshot();
    }
}
```

## Download(jar)
Version 1.02<br>
[Screenshot.jar](Screenshot.jar)

## Known bugs/issues
You can find [here](https://github.com/NmdOfficial/Screenshot/issues) the known bugs/issues

## Changelog
Watch [here](Changelog.md) the version changes

## Screenshots
![animation](https://github.com/NmdOfficial/Screenshot/blob/master/Images/Animation.gif)

## License
```
Copyright 2018 Author @NMD [Next Mobile Development]

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
```
## Last words
If you like this library feel free to "star" it:<br>
![star](https://github.com/NmdOfficial/Screenshot/blob/master/Images/star.png)

Offical E-MAIL:<br>
[nmdofficialhelp@gmail.com](mailto:nmdofficialhelp@gmail.com)


This library has been successfully tested with the latest version of Android Studio.


Changelog.md
# Changelog

| Version | Date | Changes |
| --- | --- | --- |
| `1.00` | 6th March 2018 | Initial version released |
| `1.01` | 7th March 2018 | New version released<br>- Added "Bitmap bitmap" to OnResultListener<br>- Added setDimAmount(float amount)<br>- Added getDimAmount()<br>- Added TakeScreenshotFromView()<br><br>- Fixed wrong preview dialog resolution in landscape mode |
| `1.02` | 8th March 2018 | New version released<br>- Added arePermissionsGranted()<br>- Added isReadPermissionGranted()<br>- Added isWritePermissionGranted()<br>- Added shutter click sound when creating screenshot (Android 4.1+) |
| `1.03` | 24th April 2018 | Fixed notifications bug on android 8 devices |


Screenshot.java
/*
 * Copyright 2018 Author @NMD [Next Mobile Development]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nmd.android.support;


import android.app.Activity;
import android.app.AlertDialog;
import android.app.PendingIntent;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;

import android.graphics.Bitmap;

import android.media.MediaActionSound;
import android.media.MediaScannerConnection;

import android.os.AsyncTask;
import android.os.Environment;
import android.os.Build.VERSION;
import android.os.Handler;

import android.util.DisplayMetrics;
import android.util.Log;

import android.view.Display;
import android.view.View;
import android.view.Window;
import android.view.WindowManager;

import android.widget.LinearLayout;
import android.widget.ImageView;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileOutputStream;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.lang.System;
import java.util.Timer;
import java.util.TimerTask;

import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationCompat.BigPictureStyle;
import android.support.v4.app.NotificationCompat.Builder;
import android.support.v4.app.NotificationManagerCompat;


@SuppressWarnings("unused")
public class Screenshot {

    private Activity activity;
    private Context context;
    private static final String LOG_TAG = "Screenshot";
    private static final String CHANNEL_ID = "Screenshot_Channel_ID";
    private Bitmap bitmapBackup;
    private LinearLayout layout;
    private ImageView imageView;
    private MediaActionSound takeSound;
    private OnResultListener onResultListener;
    private final Handler androidUIHandler = new Handler();
    private NotificationManagerCompat mNotificationManagerCompat;
    private PackageManager packageManager;
    
    private String fileName;
    private boolean preview;
    private boolean notification;
    private String notificationTitle;
    private String notificationShareTitle;
    private boolean notificationBigStyle;
    private boolean notificationButton;
    private String filePathBackup;
    private String fileNameBackup;
    private float dimAmount;
    private String currentPackageName;
    private int readPermission;
    private int writePermission;

    
    public Screenshot(Context context) {         
      this.context = context;
      this.activity = (Activity)context;
      Initialize();
    }
    
    private void Initialize() {
      this.fileName = "Screenshot.png";
      this.preview = true;
      this.notification = true;
      this.notificationTitle = "Screenshot..";
      this.notificationShareTitle = "Share";
      this.notificationBigStyle = false;
      this.notificationButton = true;
      this.dimAmount = 0.5f;
      this.mNotificationManagerCompat = NotificationManagerCompat.from(this.context);
      this.packageManager = this.context.getPackageManager();
      this.currentPackageName = this.context.getPackageName().toString();
      this.readPermission = packageManager.checkPermission("android.permission.READ_EXTERNAL_STORAGE", this.currentPackageName);
      this.writePermission = packageManager.checkPermission("android.permission.WRITE_EXTERNAL_STORAGE", this.currentPackageName);
          if (VERSION.SDK_INT >= 16) {
              takeSound = new MediaActionSound();
              takeSound.load(MediaActionSound.SHUTTER_CLICK);
          }
      Log.d(LOG_TAG,"Screenshot Created");
    }

    public void TakeScreenshot() {
      View view = activity.getWindow().getDecorView().getRootView();
      Take(view);
    }
    
    public void TakeScreenshotFromView(View view) {
      Take(view);
    }
    
    private void Take(View view) {
        final Timer t = new Timer();
        t.schedule(new TimerTask() {
            public void run() {
            	androidUIHandler.post(new Runnable() {
                    public void run() {
                    view.setDrawingCacheEnabled(true);
                    view.buildDrawingCache(true);
                    Bitmap bitmap = Bitmap.createBitmap(view.getDrawingCache());
                    bitmapBackup = bitmap;
                    view.setDrawingCacheEnabled(false);
                        
                    SaveUtil(bitmap);
                    MediaScannerConnection.scanFile(context, new String[] { filePathBackup }, new String[] { "image/*" }, null);
                    t.cancel();
                    }
                }); 
            }
        }, 75);
    }
    
    public boolean arePermissionsGranted() {
      return (this.readPermission == PackageManager.PERMISSION_GRANTED &&
              this.writePermission == PackageManager.PERMISSION_GRANTED);
    }
    
    public boolean isReadPermissionGranted() {
      return (this.readPermission == PackageManager.PERMISSION_GRANTED);
    }
    
    public boolean isWritePermissionGranted() {
      return (writePermission == PackageManager.PERMISSION_GRANTED);
    }
    
    public interface OnResultListener {
      public void result(boolean success, String filePath, Bitmap bitmap);
    }
   
    public void setCallback(OnResultListener listener) {
      this.onResultListener = listener;
    }
    
    public void setFileName(String name) {
      this.fileName = name;
    }

    public String getFileName() {
      return this.fileName;
    }

    private void SaveUtil(Bitmap bmOut) {
    ByteArrayOutputStream ostream = new ByteArrayOutputStream();
    bmOut.compress(Bitmap.CompressFormat.PNG, 0 , ostream);
    
    FileOutputStream fostream = null;
    File image = null;
        try {
            image = new File(Environment.getExternalStorageDirectory()+"/"+fileName);
            fostream = new FileOutputStream(image);
            fostream.write(ostream.toByteArray());
            fostream.flush();
            fostream.close();
        } catch (FileNotFoundException e) {
            Log.e(LOG_TAG, e.getMessage());
            if (onResultListener != null) {
                onResultListener.result(false, e.getMessage(), null);
            }
        } catch (IOException e) {
            Log.e(LOG_TAG, e.getMessage());
            if (onResultListener != null) {
                onResultListener.result(false, e.getMessage(), null);
            }
        }
    filePathBackup = (image != null) ? image.getAbsolutePath() : "There was a problem";
    fileNameBackup = (image != null) ? image.getName() : "There was a problem";
    //image is null should never happen
        if (VERSION.SDK_INT >= 16) {
            takeSound.play(MediaActionSound.SHUTTER_CLICK);
        }
        if (this.preview) {
            Preview();
        }
        if (this.notification) {
            NewNotification();
        }  
        if (onResultListener != null) {
            onResultListener.result(true, filePathBackup, bitmapBackup);
        }
    }

    public void ShowPreview(boolean enabled) {
      this.preview = enabled;
    }
	
    public boolean ShowPreview() {
      return this.preview;
    }
	
    public void ShowNotification(boolean enabled) {
      this.notification = enabled;
    }
	
    public boolean ShowNotification() {
    return this.notification;
    }
	
    public void NotificationTitle(String title){
      this.notificationTitle = title;   	
    }
	
    public String NotificationTitle() {
      return this.notificationTitle;
    }
	
    public void NotificationShareTitle(String title){
      this.notificationShareTitle = title;   	
    }
	
    public String NotificationShareTitle() {
      return this.notificationShareTitle;
    }
	
    public void NotificationBigStyle(boolean enabled){
      this.notificationBigStyle = enabled;   	
    }
	
    public boolean NotificationBigStyle() {
      return this.notificationBigStyle;
    }
    
    public void NotificationShareButton(boolean enabled){
      this.notificationButton = enabled;   	
    }
    	
    public boolean NotificationShareButton() {
      return this.notificationButton;
    }
	
    public void AllowScreenshots(boolean enabled){
        if (enabled) {
            activity.getWindow().clearFlags(WindowManager.LayoutParams.FLAG_SECURE);
        } else {
            activity.getWindow().setFlags(WindowManager.LayoutParams.FLAG_SECURE, WindowManager.LayoutParams.FLAG_SECURE);
        }
    }

    public void setDimAmount(float amount) {
      this.dimAmount = amount;
    }
    
    public float getDimAmount() {
      return this.dimAmount;
    }
    
    private void Preview() {
    AlertDialog.Builder alert = new AlertDialog.Builder(activity);

    layout = new LinearLayout(context);
    imageView = new ImageView(activity);
    layout.setOrientation(LinearLayout.VERTICAL);
    Display display = ((WindowManager) context.getSystemService(Context.WINDOW_SERVICE)).getDefaultDisplay();
        if (display.getRotation() == 0 || display.getRotation() == 2) {
            //for portrait mode
        	layout.setPadding(15, 19, 15, 19);
        } else {
        	//for landscape mode
            layout.setPadding(27, 0, 27, 0);
        }
    layout.setBackgroundColor(-1);//white
    imageView.setImageBitmap(bitmapBackup);
	
    layout.addView(imageView);
    alert.setView(layout);
    alert.setCancelable(false);
	
    final AlertDialog dialog = alert.create();
    dialog.getWindow().setDimAmount(this.dimAmount);
        
    dialog.show();
	
    DisplayMetrics metrics = new DisplayMetrics();
    activity.getWindowManager().getDefaultDisplay().getMetrics(metrics);

    int height = (int) (metrics.heightPixels*0.86);
    int width = (int) (metrics.widthPixels*0.85);

    dialog.getWindow().setLayout(width, height);
	
        final Timer t = new Timer();
        t.schedule(new TimerTask() {
            public void run() {
                if (dialog.isShowing()) {
                    dialog.dismiss();
                    t.cancel();
                }
            }
        }, 1250);
    }
    
    private void NewNotification() {
      int ID = (int) System.currentTimeMillis();
    	
      NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(context, CHANNEL_ID)
      .setContentTitle(notificationTitle)
      .setContentText(fileNameBackup)
      .setSmallIcon(android.R.drawable.ic_menu_gallery)
      .setPriority(NotificationCompat.PRIORITY_DEFAULT)
      .setVisibility(NotificationCompat.VISIBILITY_PUBLIC)
      .setChannelId(CHANNEL_ID)
      .setAutoCancel(true);
      
      if (VERSION.SDK_INT >= 16) {
          Intent shareIntent = new Intent();
          shareIntent.setAction(Intent.ACTION_SEND);
          shareIntent.putExtra(Intent.EXTRA_STREAM, android.net.Uri.parse(filePathBackup));
          shareIntent.putExtra("EXTRA_DETAILS_ID", ID);
          shareIntent.setType("image/*");

          if (notificationButton) {
              //Action buttons depend on expanded notifications, which are only available in Android 4.1 and later.
              PendingIntent detailsPendingIntent = PendingIntent.getActivity(context, ID, shareIntent, PendingIntent.FLAG_CANCEL_CURRENT);
              notificationBuilder.addAction(android.R.drawable.ic_menu_share, notificationShareTitle, detailsPendingIntent);
              //notificationBuilder.cancel(ID);
          }
      }
      
      if (notificationBigStyle) {
          notificationBuilder.setStyle(new NotificationCompat.BigPictureStyle().bigPicture(bitmapBackup));
      }
      
      Intent intent = new Intent();
      intent.setAction(Intent.ACTION_VIEW);
      intent.setDataAndType(android.net.Uri.parse(filePathBackup), "image/*");
      notificationBuilder.setContentIntent(PendingIntent.getActivity(context, ID, intent, 0));
      
      mNotificationManagerCompat.notify(0, notificationBuilder.build());
    }
}